app/Pasteboard.m

```diff
@@ -0,0 +1,247 @@
+        size = INITIAL_BUFFER_CAP;
+    }
+    while (size < fit_len) size *= 2;
+
+    void *new_buf = realloc(fd_priv(fd).buffer, size);
+    if (new_buf == NULL) {
+        return 1;
+    }
+
+    fd_priv(fd).buffer = new_buf;
+    fd_priv(fd).buffer_cap = size;
+
+    return 0;
+}
+
+// buffer => UIPasteboard
+static int clipboard_wsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer == NULL) {
+        return 0;
+    }
+
+    void *data = fd_priv(fd).buffer;
+    size_t len = fd_priv(fd).buffer_len;
+
+    // FIXME(stek29): This logs "Returning local object of class NSString"
+    // and I have no idea why (or how to fix it)
+    UIPasteboard.generalPasteboard.string = [[NSString alloc]
+                                             initWithBytes:data
+                                             length:len
+                                             encoding:NSUTF8StringEncoding];
+
+    // Reset generation since we've just updated UIPasteboard
+    // note: offset doesn't change
+    fd_priv(fd).generation = 0;
+
+    return 0;
+}
+
+// UIPasteboard => buffer, return len
+static ssize_t clipboard_rsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer != NULL) {
+        free(fd_priv(fd).buffer);
+        fd_priv(fd).buffer = NULL;
+        fd_priv(fd).buffer_cap = 0;
+        fd_priv(fd).buffer_len = 0;
+    }
+
+    size_t len;
+    const void *data = get_data(fd, &len);
+    
+    // Make sure size is still INITIAL_BUFFER_CAP based
+    if (realloc_to_fit(fd, len)) {
+        return _ENOMEM;
+    }
+
+    memcpy(fd_priv(fd).buffer, data, len);
+    fd_priv(fd).buffer_len = len;
+
+    return len;
+}
+
+static int clipboard_poll(clip_fd *fd) {
+    int rv = POLL_WRITE;
+
+    size_t len = 0;
+    if (get_data(fd, &len) != NULL && fd->offset < len) {
+        rv |= POLL_READ;
+    }
+
+    return rv;
+}
+
+static ssize_t clipboard_read(clip_fd *fd, void *buf, size_t bufsize) {
+    size_t length = 0;
+    const char *data = get_data(fd, &length);
+
+    if (data == NULL) {
+        return _EIO;
+    }
+
+    size_t remaining = length - fd->offset;
+    if ((size_t) fd->offset > length)
+        remaining = 0;
+
+    size_t n = bufsize;
+    if (n > remaining)
+        n = remaining;
+
+    if (n != 0) {
+        memcpy(buf, data + fd->offset, n);
+        fd->offset += n;
+    }
+
+    return n;
+}
+
+static ssize_t clipboard_write(clip_fd *fd, const void *buf, size_t bufsize) {
+    size_t new_len = fd->offset + bufsize;
+    size_t old_len = fd_priv(fd).buffer_len;
+
+    if (realloc_to_fit(fd, new_len)) {
+        return _ENOMEM;
+    }
+
+    // fill the hole between new offset and old len
+    if (old_len < fd->offset) {
+        memset(fd_priv(fd).buffer + old_len, '\0', fd->offset - old_len);
+    }
+
+    memcpy(fd_priv(fd).buffer + fd->offset, buf, bufsize);
+    fd->offset += bufsize;
+    fd_priv(fd).buffer_len = new_len;
+
+    return bufsize;
+}
+
+static off_t_ clipboard_lseek(clip_fd *fd, off_t_ off, int whence) {
```

Comment:
Not something you need to fix in this PR, but this logic is duplicated in procfs, so it should be moved to a utility function at some point.